File: 0216-Combination_Sum_III/combination_sum_iii.go
package _216_Combination_Sum_III

var (
	result [][]int
	path   []int
)

func combinationSum3(k int, n int) [][]int {
	result, path = make([][]int, 0), make([]int, 0, k)
	backtracking(k, n, 0, 1)
	return result
}

func backtracking(k int, n int, sum int, startIndex int) {
	if sum > n {
		return
	}
	if len(path) == k {
		if sum == n {
			tmp := make([]int, k)
			copy(tmp, path)
			result = append(result, tmp)
		}
		return
	}

	for i := startIndex; i <= 9 - (k - len(path)) + 1; i++ {
		sum += i
		path = append(path, i)
		backtracking(k, n, sum, i+1)
		sum -= i
		path = path[:len(path)-1]
	}
}

File: leetcode/0216-Combination_Sum_III/combination_sum_iii.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
private:
    vector<vector<int>> result;
    vector<int> temp_set;
    void backtracking(int k, int n, int sum, int startIndex)
    {
        if (temp_set.size() == k)
        {
            if (sum == n) result.push_back(temp_set);
            return;
        }
        for (int i = startIndex; i <= 9; i++)
        {
            sum += i;
            temp_set.push_back(i);
            backtracking(k, n, sum, i + 1);
            sum -= i;
            temp_set.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        result.clear();
        temp_set.clear();
        backtracking(k, n, 0, 1);
        return result;
    }
};

void show_result(vector<vector<int>> result)
{
    cout << "[";
    for (const vector<int>& res : result)
    {
        cout << "[";
        for (const int i : res)
        {
            cout << i << " ";
        }
        cout << "], ";
    }
    cout << "]" << endl;
}

int main(int argc, char* argv[])
{
    Solution sol;
    show_result(sol.combinationSum3(3, 7));
    show_result(sol.combinationSum3(3, 9));
    show_result(sol.combinationSum3(4, 1));

    return 0;
}
